File: JavaScript/5_DataStructures/1_SinglyLinkedList.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

class SingleLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }
  printAll() {
    // Just For Seeing , not Efficient
    if (!this.head) return -1;
    let node = this.head;
    let output = [];
    while (node) {
      output.push(node.val);
      node = node.next;
    }
    return output;
  }
  push(val) {
    //Append new val to the end of the list
    const new_node = new Node(val);
    if (!this.head) {
      this.head = new_node;
      this.tail = new_node;
    } else {
      this.tail.next = new_node;
      this.tail = new_node;
    }
    this.length++;
    return true;
  }
  pop() {
    //pop the last value of the list
    if (!this.head) return false;
    if (this.length === 1) {
      let node = this.head;
      this.head = null;
      this.tail = null;
      this.length = 0;
      return node.val;
    }
    let node = this.head;
    let previousNode = node;
    while (node !== this.tail) {
      previousNode = node;
      node = node.next;
    }
    previousNode.next = null;
    this.tail = previousNode;
    this.length--;
    return node.val;
  }

  shift(val) {
    //add new element to the Starting of the list
    const new_node = new Node(val);
    if (!this.head) this.push(val);
    else {
      let oldHead = this.head;
      new_node.next = oldHead;
      this.head = new_node;
      this.length++;
      return true;
    }
  }

  unShift() {
    //Removes the 1st element from the list
    if (this.length < 1) this.pop();
    else {
      let oldHead = this.head;
      this.head = oldHead.next;
      oldHead.next = null;
      this.length--;
      return oldHead;
    }
  }

  get(index) {
    //Get the Node at Index "index"
    if (index < 0 || index >= this.length) return false;
    let node = this.head;
    for (let i = 1; i <= index; i++) {
      node = node.next;
    }
    return node;
  }

  set(val, index) {
    let setNode = this.get(index);
    if (!setNode) {
      return false;
    }
    setNode.val = val;
    return true;
  }

  insert(val, index) {
    if (index < 0 || index > this.length) return false;
    if (index === 0) return this.shift(val);
    if (index === this.length) return this.push(val);
    const new_node = new Node(val);
    let connectNode = this.get(index - 1);
    new_node.next = connectNode.next;
    connectNode.next = new_node;
    this.length++;
    return true;
  }

  remove(index) {
    if (index < 0 || index >= this.length) return false;
    if (index === 0) return this.unShift();
    if (index === this.length - 1) return !!this.pop();
    let removeNodeParent = this.get(index - 1);
    removeNodeParent.next = removeNodeParent.next.next;
    this.length--;
    return true;
  }
  reverse() {
    if (!this.head) return false;
    let node = this.head;
    this.head = this.tail;
    this.tail = node;
    let next;
    let prev = null;
    for (let i = 0; i < this.length; i++) {
      next = node.next;
      node.next = prev;
      prev = node;
      node = next;
    }
    return this;
  }
}

let l = new SingleLinkedList();

File: Python/5_DataStructures/1_SinglyLinkedList.py
'''
for a given single linked List check if there is any cycle in between them
'''

class Node(object):
	def __init__(self, val):
		self.val = val
		self.next = None



class LinkedList(object):
	def __init__(self):
		self.head = None
		self.tail = None
		self.length = 0

	def printAll(self):
		if !self.head: return False

		elements = []
		node = self.head

		while node:
			elements.append(node.val)
			node = node.next

		return elements

	def push(self, val):
		new_node = Node(val)

		if !self.head:
			self.head = self.tail = new_node
		else:
			self.tail.next = new_node
			self.tail = new_node

		self.length += 1

		return new_node.val

	def pop(self):
		if self.length == 1:
			node = self.head
			self.head = self.tail = None
		else:
			node = self.head
			previous_node = node

			while node != self.tail:
				previous_node = node
				node = node.next

			previous_node.next = None
			this.tail = previous_node

		this.length -= 1

		return node.val


	def unshift(self, val):
		new_node = Node(val)

		if !self.head:
			self.head = self.tail = new_node
		else:
			new_node.next = self.head
			self.head = new_node
		self.length += 1 

		return new_node.val


	def shift(self):
		if self.length == 1:
			node = self.head
			self.head = self.tail = None
		else:
			node = self.head
			self.head = self.head.next
		
		return node.val


	def get(self, index):
		if index > self.length or index < 0: return False

		node = self.head
		for i in range(index):
			node = node.next

		return node

	def set(self, index, val):
		set_node = self.get(index)

		if !set_node: return False

		set_node.val = val

		return set_node


	def insert(self, index, val):
		if index < 0 or index > self.length: return False
		if index == 0: self.unshift(val)
		if index == self.length: self.push(val)

		prev_node = self.get(index-1)
		new_node = Node(val)
		
		new_node.next = prev_node.next
		prev_node.next = new_node
		self.length += 1

		return True


	def remove(self, index):
		if index < 0 or index > self.length-1: return False
		if index == 0: self.shift()
		if index == self.length-1: self.pop()

		prev_node = self.get(index-1)

		prev_node.next = prev_node.next.next
		self.length -= 1
		return True


	def reverse(self):
		pass

	




